implementation/src/main/java/io/smallrye/mutiny/operators/UniCache.java

```diff
@@ -70,17 +73,42 @@ private synchronized void onCancellation(UniSubscriber<? super I> subscriber) {
         subscribers.remove(subscriber);
     }
 
-    private void replay(UniSubscriber<? super I> subscriber) {
-        synchronized (this) {
-            if (state != COMPLETED) {
-                throw new IllegalStateException(
-                        "Invalid state - expected being in the DONE state, but is in state: " + state);
-            }
-        }
-        if (failure != null) {
-            subscriber.onFailure(failure);
+    /**
+     * Handling a new subscriber this UniCache is already {@link #COMPLETED}.
+     * Verifies the cached result against the {@link #validator} and
+     * either replays the cached result or renews our own subscription
+     * to restart the cache lifecycle.
+     *
+     * @param subscriber the newly subscribed {@link UniSubscriber}
+     * @return the {@link Runnable} to be executed afterwards
+     */
+    private Runnable handleCompleted(UniSubscriber<? super I> subscriber) {
+        subscribers.add(subscriber);
+
+        // Closure of item and failure for validity checking and replaying the same objects
+        I item = this.item;
+        Throwable failure = this.failure;
+
+        // Test whether the cache is still valid
+        // Done within sync-block to avoid multiple subscriptions to the upstream
+        if (validator.test(item, failure)) {
```

Comment:
Not quite sure about the call of an external provided callback (the predicate) within the synchronized block.
Had no better idea until now, as the result (`.test() == false`) may cause an imediate state change.

Splitting up the synchronized or synchronizing again in the `action` runnable and working with closures of item and failure could be an option.
Or may it be sufficient to clearly state out in the documentation that the predicate must be implemented pure functional?
Your opinion?

Reply:
That could be tricky.

What about using a separate lock object that guarantees the validator is not called concurrently and avoid calling the validator with the monitor lock. 

The separate lock should not be reachable from the validator code. 
However, I would also mention in the documentation that the validator should not have side-effects.

Reply:
Ok, thank you.
I've reworked the COMPLETED action, tests are running unchanged.
Hope it comes near to what you had in mind.